lib/charms/saml_integrator/v0/saml.py

```diff
@@ -0,0 +1,303 @@
+        result[f"{prefix}url"] = self.url
+        result[f"{prefix}binding"] = self.binding
+        if self.response_url:
+            result[f"{prefix}response_url"] = self.response_url
+        return result
+
+    @classmethod
+    def from_relation_data(cls, relation_data: Dict[str, str]) -> "SamlEndpoint":
+        """Initialize a new instance of the SamlEndpoint class from the relation data.
+
+        Args:
+            relation_data: the relation data.
+
+        Returns:
+            A SamlEndpoint instance.
+        """
+        url_key = ""
+        for key in relation_data:
+            # A key per method and entpoint type that is always present
+            if key.endswith("_redirect_url") or key.endswith("_post_url"):
+                url_key = key
+        # Get endpoint name from the relation data key
+        lowercase_name = "_".join(url_key.split("_")[:-2])
+        name = "".join(x.capitalize() for x in lowercase_name.split("_"))
+        # Get HTTP method from the relation data key
+        http_method = url_key.split("_")[-2]
+        prefix = f"{lowercase_name}_{http_method}_"
+        return cls(
+            name=name,
+            url=parse_obj_as(AnyHttpUrl, relation_data[f"{prefix}url"]),
+            binding=relation_data[f"{prefix}binding"],
+            response_url=parse_obj_as(AnyHttpUrl, relation_data[f"{prefix}response_url"])
+            if f"{prefix}response_url" in relation_data
+            else None,
+        )
+
+
+class SamlRelationData(BaseModel):
+    """Represent the relation data.
+
+    Attrs:
+        entity_id: SAML entity ID.
+        metadata_url: URL to the metadata.
+        certificates: List of SAML certificates.
+        endpoints: List of SAML endpoints.
+    """
+
+    entity_id: str = Field(..., min_length=1)
+    metadata_url: AnyHttpUrl
+    certificates: List[str]
+    endpoints: List[SamlEndpoint]
+
+    def to_relation_data(self) -> Dict[str, str]:
+        """Convert an instance of SamlDataAvailableEvent to the relation representation.
+
+        Returns:
+            Dict containing the representation.
+        """
+        result = {
+            "entity_id": self.entity_id,
+            "metadata_url": self.metadata_url,
+            "x509certs": ",".join(self.certificates),
+        }
+        for endpoint in self.endpoints:
+            result.update(endpoint.to_relation_data())
+        return result
+
+
+class SamlDataAvailableEvent(ops.RelationEvent):
+    """Saml event emitted when relation data has changed.
+
+    Attrs:
+        entity_id: SAML entity ID.
+        metadata_url: URL to the metadata.
+        certificates: Tuple containing the SAML certificates.
+        endpoints: Tuple containing the SAML endpoints.
+    """
+
+    @property
+    def entity_id(self) -> str:
+        """Fetch the SAML entity ID from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("entity_id")
```

Comment:
Is `self.relation.data[self.relation.app]` always there ? There is no chance that the relation get created without any data in the relation bag initially ?

Reply:
The event is fired on relation changed and the information populated on relation created, so yes, it should be present